File: src/web/DuckDuckGoImageQBuilder.java
package web;

import java.net.URI;
import util.parsing.ParsesFromString;
import util.parsing.StringParseStrategy;
import util.parsing.StringParseStrategy.From;
import util.plugin.IsPlugin;
import static util.Util.urlEncodeUtf8;
import static util.parsing.StringParseStrategy.To.CONSTANT;

/**
 * @author Martin Polakovic
 */
@IsPlugin
@StringParseStrategy(from = From.ANNOTATED_METHOD, to = CONSTANT, constant = "DuckDuckGo")
public class DuckDuckGoImageQBuilder implements SearchUriBuilder {

	@ParsesFromString
	public DuckDuckGoImageQBuilder() {}

	@Override
	public URI apply(String q) {
		return URI.create("https://duckduckgo.com/?q=" + urlEncodeUtf8(q));
	}

}

File: src/web/DuckDuckGoImageQBuilder.kt
package web

import util.parsing.StringParseStrategy
import util.parsing.StringParseStrategy.From
import util.parsing.StringParseStrategy.To
import java.net.URI

@StringParseStrategy(from = From.SINGLETON, to = To.CONSTANT, constant = "DuckDuckGo Image")
object DuckDuckGoImageQBuilder: ImageSearchUriBuilder {

    override fun doApply(q: String): URI {
        return URI.create("https://duckduckgo.com/?q=$q")!!
    }

}